### ISSUE TITLE
removed from plugin manager

### ISSUE THREAD

**Issue by psifertex** [fluxchief/binaryninja_avr#15]
Apologies for the late notice, we removed this from the plugin manager some time last year due to:

https://github.com/fluxchief/binaryninja_avr/blob/master/__init__.py#L416-L418

I'd love to re-add it back once that is resolved. Very sorry for not notifying you directly earlier (don't remember if we talked via another mechanism or not)

**Comment by fluxchief** [comment]
You actually mentioned this in [issue #14](https://github.com/fluxchief/binaryninja_avr/issues/14#issuecomment-902094549), so this was no surprise.
As explained in the bug, fixing this block is actually a bit challenging there are no magic bytes which can be used to differentiate between a valid AVR binary or just a random blob.
It might be possible to work around this with some heuristics, however that's where the second issue comes into play: different AVR chips have different amounts of ISRs and rom sizes. A heuristic that checks that ISRs actually point into the image needs to know how many ISRs exists for the specific chip.
A better solution would be to have to explicitly open the bin with the architecture plugin and having to select the corresponding chip.
Realistically I still don't see myself fixing this anytime soon, haven't been doing lots of RE work lately (and don't even have a up2date BN)

Closing this as dupe of #14

**Comment by fluxchief** [comment]
On a second thought let's keep it open so that it's clear to others why this is not part of the plugin repo

**Comment by fluxchief** [comment]
With the "linear sweep" pass added to BN, creating the ISR routines is not as important anymore - that code is currently disabled.
The "apply BV during open-with-options" would work, but ideally we'd figure out why it is locking up to begin with.

**Comment by psifertex** [comment]
So I think the link I had above is incorrect due to changes in the file. The problem isn't adding the ISR functions but the `is_valid` always returning true which means this view gets called _every time_ and not only slows down analysis for all other files but means any bug in the lifting or analysis can cause problems even with files that aren't AVR. That looks to still be true and would have to change before it can be re-added. 

You can also register as an ELF handler even so that elf based on AVR will still automatically load like:
https://github.com/Vector35/arch-mips/blob/master/arch_mips.cpp#L2136-L2139

Basically, just call [register_arch](https://api.binary.ninja/binaryninja.binaryview-module.html#binaryninja.binaryview.BinaryViewType.register_arch) with the appropriate ident.

That said, yeah, I don't think we have a good way to let the user override which BinaryViews are loaded for a given file, but it sounds like something that would be a good idea and easy enough to do. Lemme talk to the team and see if there's any concerns.

**Comment by psifertex** [comment]
I believe this to be unblocked now with the resolution of #3721.

Happy to submit a PR, but the TL;DR is implement `is_force_loadable` and have it return true in your BinaryView. This will prevent it from applying to _all_ files but let the user specify they'd like to load it using "open with options".

**Comment by fluxchief** [comment]
Awesome, thanks for letting me know. I'll have a look but probably won't be able to do fix this in the next week. Happy to accept your PR if you feel like sending one before that.

**Comment by psifertex** [comment]
Totally fine, no rush on that.

**Comment by Grazfather** [comment]
Is this literally all that's needed?

```diff
diff --git i/__init__.py w/__init__.py
index e7529d9..bd570cf 100644
--- i/__init__.py
+++ w/__init__.py
@@ -436,6 +436,9 @@ class AVRBinaryView(binaryninja.BinaryView):
     def is_valid_for_data(self, data):
         return True

+    def is_force_loadable(self):
+        return True
+
     @classmethod
     def get_load_settings_for_data(cls, data):
         load_settings = binaryninja.Settings("avr")
```

I have had this plugin installed for years, but I basically moved the clone around to disable it to avoid the aforementioned issues. It would be nice if I didn't have to do this.

**Comment by psifertex** [comment]
That's correct. Along with removing the return true in the is_valid. 

A new release has to be generated though which only the main repo author can do. If they aren't available and you'd like to maintain a fork, be happy to add that to the plugin manager.

**Comment by fluxchief** [comment]
I tried implementing this in the past but didn't manage to get it to work within a reasonable timeframe - mostly because I was unable to navigate around the API.

> That's correct. Along with removing the return true in the is_valid.

According to the documentation this attribute (`is_force_loadable`) is for `BinaryViewType` and not for `BinaryView.` It is not mentioned in the `BinaryView` documentation. I think I tried it in the past and couldn't get it to run, but maybe I'm mistaken and it just is that simple.

I'm more than happy to accept PRs, so if you manage to get this to behave as expected, please send it my way.

**Comment by Grazfather** [comment]
I can try, but I have no idea how to get it to load. How do I get binary ninja to prompt me? It just (incorrectly) assumed this elf was an x86 binary.

```bash
❯ /opt/homebrew/Cellar/avr-gcc\@8/8.5.0_2/bin/avr-gcc -mmcu=atmega328 -Wall -Os -o test.elf test.c -static

❯ file test.elf
test.elf: ELF 32-bit LSB executable, Atmel AVR 8-bit, version 1 (SYSV), statically linked, with debug_info, not stripped
```

![Image](https://github.com/user-attachments/assets/77383bf4-1ea2-442d-a018-253e423a0672)

This is with 'both' changes:

```diff
diff --git i/__init__.py w/__init__.py
index e7529d9..152ca6c 100644
--- i/__init__.py
+++ w/__init__.py
@@ -434,6 +434,9 @@ class AVRBinaryView(binaryninja.BinaryView):

     @classmethod
     def is_valid_for_data(self, data):
+        return False
+
+    def is_force_loadable(self):
         return True

     @classmethod
```

**Comment by psifertex** [comment]
FYI, I've got some code in testing that should resolve most of this (afaict anyway) at: https://github.com/psifertex/binaryninja_avr

I'll PR it as soon as I have a bit more confidence in testing. :-)
